### PR TITLE
provide opt to disable mime-type security warning

### PR DESCRIPTION
**Pull request by jrochkind**
Shrine.opts[:mime_type_security_warning] = :suppress

Used in shrine test env now instead of monkey patch that disabled all warnings

### CODE REVIEW

File: lib/shrine.rb [hunk 1]
```diff
@@ -295,7 +295,9 @@ def extract_filename(io)
         # Attempts to extract the MIME type from the IO object.
         def extract_mime_type(io)
           if io.respond_to?(:content_type) && io.content_type
-            warn "The \"mime_type\" Shrine metadata field will be set from the \"Content-Type\" request header, which might not hold the actual MIME type of the file. It is recommended to load the determine_mime_type plugin which determines MIME type from file content."
+            if opts[:mime_type_security_warning] != :suppress
```
Comment (hmistry) [inline lib/shrine.rb]: I prefer a simpler name like `:mime_type_warning`.

On a higher level, do we want to go down the path of having individual warning control or a global warning control?

Comment (jrochkind) [reply lib/shrine.rb]: I can change the name to whatever y'all like. 

I think individual control is needed. I thought about this in the first place because I specifically did _not_ want to disable _all_ warnings (including some I dont' even know about). 

I think the shrine source code probably doesn't want to either -- it does a monkey-patch to disable all warnings, next to a comment saying it was only intending to disable a certain warning. If the shrine tests produced _other_ warnings that were intended to be produced by shrine code (they do not at present), I don't think we'd want em hidden. 

The monkey-patch approach is still available for disabling all warnings, which is all that was available before anyway.

Comment (jrochkind) [reply lib/shrine.rb]: This is the only warning there is a previous known use case for disabling -- I just want it to be possible to disable just that warning, which is what I think was desired, and not via a monkey-patch is bonus too.

Comment (hmistry) [reply lib/shrine.rb]: So you prefer individual warning control - ok. Me too, for now.

I only raise the question for us to make a proactive decision on warnings going forward.

**Comment by janko-m** [comment]
Thanks, but it already should be possible to avoid the warning by loading the `determine_mime_type` plugin with the `:default` analyzer. At least that was the intention, but a few days ago I realized that the warning is still being printed by accident. So I [fixed](https://github.com/shrinerb/shrine/commit/edd28f6abbe1878cdef91760cc740b86120664ee) it in master. I also renamed the `:default` analyzer to `:content_type` to be more explicit.

TL;DR: On `master` you can avoid the warning:

```rb
plugin :determine_mime_type, analyzer: :content_type
```

**Comment by jrochkind** [comment]
@janko-m 

Can you comment on this part of shrine's test setup?

https://github.com/shrinerb/shrine/blob/203c2c9a0c83815c9ded1b09d5d006b2a523579c/test/test_helper.rb#L20-L22

It is monkey-patching to remove ALL warnings, with a comment that says "disable mime_type warnings". 

Is it okay that it's removing all warnings, might we want to see warnings in tests that _aren't_ the mime-type warning, if they are accidentally triggered?

Should it have a way to do this other than monkey-patching?

Is there a better way for it to be doing this, should the test helper instead add `plugin :determine_mime_type, analyzer: :content_type`?

I am writing tests in my own code of custom-to-me shrine functionality. I looked to how shrine did testing to copy it. That is when I encountered this, and thought, gee, there should be a (better) way to disable just the mime-type warning and not all warnings. 

No?

**Comment by janko-m** [comment]
@jrochkind Yep, I definitely agree that monkey patching `warn` in tests isn't ideal, and you're right that we do want to remove warning only for the MIME type. However, I also wouldn't like to add a Shrine option that would be used only in tests.

Note that `Shrine#warn` (instance method) is only called in `Shrine#extract_mime_type` in `lib/shrine.rb` as far as I can see. Every other warning is called through `Shrine.deprecation`, which calls `Shrine.warn` (class method).

I'm not opposed to having options like these that aren't related to plugins, but I feel like it's redundant in this case, because of the alternative and the current workaround working ok (though not ideal, I admit).

**Comment by janko-m** [comment]
> They'll still be silenced by the override of warn though, right?

You'll notice that a few places in Shrine test suite are testing deprecated behaviour, and there a `deprecated` block, which is a wrapper around `it` but with silencing the deprecation warning.

https://github.com/shrinerb/shrine/blob/master/test/support/deprecated_helper.rb

Try replacing one of those blocks with `it` and you'll see that you get deprecation warning. Overriding `Shrine#warn` like we do in tests won't override `Shrine.warn` or any other `warn` methods.

> But should I just use the warn patch for now? Do you have any ideas about what a less non-ideal solution might look like?

I want your solution to sit a bit longer, and I'll get back to you. The solution of overriding `Shrine#warn` is only not ideal in the sense that potentially in the future we might start calling `Shrine#warn` somewhere more, but I doubt that will ever happen.

**Comment by hmistry** [comment]
@janko-m I can accept your "not necessary to add option for tests" argument (especially since it's our internal tests). However what about for production logs, would dev's prefer to have the warning or not? I don't recall seeing any support issues but that's where I'd base the major decision on.

You also made a good point on having warning options spread amongst plugins if we go down this path - and that doesn't sound nice. There's pros and cons either way. Maybe an alternate option is having a `warning_level` option in `Shrine::Base` which can be set depending on environment and the plugins use it accordingly.

My two cents... I'll let you sleep on it. 😃

**Comment by janko-m** [comment]
> However what about for production logs, would dev's prefer to have the warning or not? I don't recall seeing any support issues but that's where I'd base the major decision on.

Oh, I see, you think we should add a dedicated Shrine logger with configurable output? Honestly, I think that's an overkill for only this and the deprecation warnings. Both the mime type warning and deprecation warnings are not something that will be logged in a certain case, they will always be logged if you upload a file without loading `determine_mime_type` plugin or you're using an option or plugin that's deprecated. Even ActiveSupport just prints to stderr if `Rails.logger` is not set (which essentially what `warn` does).

I don't currently see any problem with the MIME type security warning and how we're silencing it in tests.

**Comment by janko-m** [comment]
Since the current `Shrine#warn` monkey patch silences only the MIME type warning, this option is redundant. This monkey patch is only needed in Shrine tests, other libraries and their test suites can set

```rb
Shrine.plugin :determine_mime_type, analyzer: :default # 2.13.0 or earlier
# or
Shrine.plugin :determine_mime_type, analyzer: :content_type # current master
```

to get rid of the warning. For me the small benefit of a bit cleaner Shrine setup doesn't justify adding a new option.

**Comment by jrochkind** [comment]
No prob, that makes sense. 

> This monkey patch is only needed in Shrine tests, other libraries and their test suites can…

But I'm going to keep using the patch approach in my own tests, because I want to be sure I'm testing my plugins in isolation, for the same reason shrine wants to.  But that works.
